### ISSUE TITLE
Plugin is very slow

### ISSUE THREAD

**Issue by SebT** [EddyVerbruggen/SocialSharing-PhoneGap-Plugin#418]
Hi,

First, thank you for this plugin.

I'm using it on Android (latest cordova version, testing on Lollipop). The problem is that the share dialog is very slow to show. Here's my code:
```javascript
// Click handler
console.log("shareOffer");
var
  message = "Offre d'emploi " + this.state.title,
  subject = "Offre d'emploi " + this.state.title,
  image   = null,
  link    = "http://www.google.com/" + pathToAppend;
window.plugins.socialsharing.share(message, subject, image, link);
```

The console instantly logs 'shareOffer' on click, but the modal shows 2-4 seconds afterwards. I noticed that if i close and reopen the dialog, it opens quickly enough. Only the first opening is slow, which kills the UX :(

I saw others had the problem (i.e #382) but I have 0 errors in the webview and 0 in logcat.

**Comment by EddyVerbruggen** [comment]
Hi,

Does it make a difference if you first call `available` or `canShareVia`? I mean just for science, if you silently call those, and later call `share`?

**Comment by SebT** [comment]
Well I tried: the callbacks from `available` and `canShareVia` are executed almost instantly but the dialog isn't showing faster.

**Comment by EddyVerbruggen** [comment]
Thanks for trying!

A few questions:
* What is the exact URL you're sharing?
* What device are your running on (hardware, Android version)?
* How many apps show up in the share sheet (approx.) ?

**Comment by SebT** [comment]
- The app I'm building is the mobile adaptation of the recruitment website http://www.meteojob.com. I'm allowing the user to share the URL of a job offer. Example :http://www.meteojob.com/candidat/offres/offre-d-emploi-ingenieur-h-f-genas-rhone-alpes-recrutement-et-placement-cdi-3209894
- LG Spirit android 5.0.1 & OnePlus One android 5.1.1
- 20+

I've only tested with my LG Spirit until today. I noticed that it is much faster on the OnePlus One, but still a bit slow, it takes ~1sec to show. 

I guess I can't do much to boost it up, can I?

**Comment by bau720123** [comment]
+1
I have the same problem 
first time，the modal shows 2-4 seconds，but if reopen the dialog，it opens quickly enough

**Comment by gimox** [comment]
+1 same problem here

**Comment by EddyVerbruggen** [comment]
@gimox What are you trying to share? Please post the entire line of code, also on which hardware/software?

**Comment by gimox** [comment]
the first time, modal open very slowly, 2/4 seconds , but after the dialog is very quickly.
(out of this thread... if i share a image with facebook it say that need a facebook login, problem or configuration?)

**Comment by EddyVerbruggen** [comment]
Where does the img come from? Is it local, and is it large perhaps?

**Comment by gimox** [comment]
img is a link in public server.
share work perfect with twitter or email (not display in whatsapp, but no errors).

Facebook say a dialog that i need login for image sharing and close the dialog.

**Comment by EddyVerbruggen** [comment]
* Try sharing the image from the local filesystem (`www/img/whatever.png`) and see if that makes if a difference. If so, download the img first, then share it (that's what the plugin does as well). I'm thinking about showing a 'spinner' or something while the plugin is downloading files btw.
* Log on via the iOS settings app, you will find Facebook there. Or install the Fb app.

**Comment by gimox** [comment]
facebook app is installed, logged and work perfect!!!
the sharing plugin stop before loading image.

**Comment by gimox** [comment]
..i found the problem.
I'm log out from facebook app. 
I log in again but after some xcode building i found facebook slogged, very strange...

**Comment by revolunet** [comment]
the slowness is generally due to the download of the a image when you provide an image url. switch to local image or base64

**Comment by gimox** [comment]
ok thanks. it's possibile open the windows and after that loading the image?

**Comment by revolunet** [comment]
not with the current code

**Comment by gimox** [comment]
ok thanks for reply. i think this is not a big problem. Have you some tips to increase image speedy?

**Comment by flyingsky** [comment]
I think this plugin should improve to solve this issue, the image should be loaded in finally sharing, not when display it.

**Comment by EddyVerbruggen** [comment]
@flyingsky If you want to display an image preview it must be downloaded first. You could preload the image from the web and use a cached version for sharing.

**Comment by flyingsky** [comment]
I think this is the only way to do it before any improvement. What I have to do in my phonegap project like below:

1. I have canvas item in my index.html

```
<!-- this is used to generate image base64 for sharing -->
    <div style="display: none">
      <canvas id="base64Canvas"></canvas>
      <img id="base64Img" crossOrigin="anonymous"/>
    </div>
```

2. get base64 data of image. note in below code if your img.src is same to your new url, onload function will not be invoked. You can get base64 image data from canvas directly.

```
var img = document.getElementById('base64Img');
img.src = "your_image_src";
img.onload = function() {
    var c = document.getElementById('base64Canvas');
    c.width = img.width;
    c.height = img.height;
    var x = c.getContext('2d');
    x.drawImage(img, 0, 0);
    var imgBase64 = c.toDataURL();
    // now call plugin to share
    ...
}
```

**Comment by VictorBjelkholm** [comment]
Experiencing the same issue on Phonegap Android.

Have the following code:
`window.plugins.socialsharing.share(null, null, null, link)`

And `link` is just a string. Very simple and should not take 1-2 seconds for the sharing dialog to popup.

Also, when pressing the share icon, it freezes the application while doing _something_ so if you happen to have any GIF or animations in the background, those stops completely, indicating that the plugin blocks the UI thread...

**Comment by EddyVerbruggen** [comment]
Well, Android is resolving that URL (link) to see if it exists and what mimetype it has. It's not the plugin that's doing that, it's Android itself. That's also why sharing a local image is much faster than sharing a remote image; it needs to be retrieved in order to determine the mimetype and it's shared from the local filesystem.

As for threading, the plugin code runs in a separate thread as recommended [here](https://cordova.apache.org/docs/en/latest/guide/platforms/android/plugin.html) - check SocialSharing.java at `doSendIntent`. Please let me know if you see anything that can be enhanced.

**Comment by VinceOPS** [comment]
Hi everyone,

Sorry to keep the conversation going but I'm experiencing the same issue on my Galaxy A5 (2016), even sharing a simple title/message (no URL, no URI, no picture, nothing). The very first share takes ~1s (maybe a bit more) to open the global/general share.

I'm currently using `shareWithOptions` feature with the last version of the plugin (5.1.1).

Any clue? I may be doing something wrong, yet I doubt it. (I also tried to wrap the share call in a setTimeout, but it has no effects).

Thanks! (And again, thanks for your work @EddyVerbruggen ).

**Comment by vikas0vanvi** [comment]
Is there any workaround like showing loading screen till the image is downloaded and then open share dialogue?

**Comment by EddyVerbruggen** [comment]
If it's a remote img then download it before sharing and use the local file.

**Comment by dan-at-rmend** [comment]
With a remote video will the plugin download the whole video before displaying the share dialogue?

**Comment by EddyVerbruggen** [comment]
Yes. It has to. Do do it manually and feed the plugin the local path.

**Comment by jamminjames** [comment]
@EddyVerbruggen Adding a spinner would be great. This is an old thread... any chance that will be done soon? Without it, a user could get discouraged, and think it's not working.

**Comment by EddyVerbruggen** [comment]
Happy to take a PR.

**Comment by jamminjames** [comment]
I'm a novice. I hope you can add it though, it seems like it would be relatively simple for someone who knows what they're doing, yes? And, as I mentioned, important, because users can easily think it's just not working and back out before the share window pops up.

**Comment by VinceOPS** [comment]
I personally did it with HTML/CSS. Simply using ionic's $ionicBackdrop
service to show a backdrop before I call the plugin. You could use
$ionicLoading the same way.

Le 4 nov. 2017 22:11, "James Israel" <notifications@github.com> a écrit :

I'm a novice. I hope you can add it though, it seems like it would be
relatively simple for someone who knows what they're doing, yes? And, as I
mentioned, important, because users can easily think it's just not working
and back out before the share window pops up.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/EddyVerbruggen/SocialSharing-PhoneGap-Plugin/issues/418#issuecomment-341930125>,
or mute the thread

**Comment by jamminjames** [comment]
@VinceOPS, You need to be using the Ionic framework for building your app to use this, right?

**Comment by jamminjames** [comment]
I found an easy solution for now. Since mine is a PhoneGap Build app, I just added the cordova-plugin-spinnerdialog plugin in config.xml, and call it before the window.plugins.socialsharing.share line that triggers this plugin; then hide it after, with `setTimeout(function(){ window.plugins.spinnerDialog.hide(); }, 4000);`. I have to use the timeout, because even though the 'hide' line is after the 'share' line, it triggers right away.
